### PR TITLE
Dynamic tables & error logger feedback 

### PR DESCRIPTION
**Pull request by GeoDerp**
The functional implementation of https://github.com/davidusb-geek/emhass/issues/186 & https://github.com/davidusb-geek/emhass/issues/191

Changes:
- new actionLogs.txt is generated in datapath storing sessions app.logger info
- on successful html button press, fetch is called to get  html containing latest table data 
- on html button press, If app.logger ERROR is present, send action log back and present on page. 

**Note:** As of first commit, this PR still needs refining, please wait to merge. Any suggestions appreciated.

### CODE REVIEW

**Comment by GeoDerp** [comment]
@davidusb-geek feel free to test this and see what you think.

**Comment by davidusb-geek** [comment]
Those are some great functionalities!
Ok, will do some testing this week-end.

**Comment by davidusb-geek** [comment]
Ok no problem. Like I said I will only be able to test this week-end

**Comment by purcell-lab** [comment]
@GeoDerp you are doing some fantastic work in extending the functionality of EMHASS.  

Thanks for your assistance! 🎉🎉

**Comment by GeoDerp** [comment]
Two notes on the error logs side of things:
- it only prints app.logger outputs. 
  - If we continue going down this route we will just need to make sure we got app.logger.error for all likely fail cases. (If there isn't already)
   - else, we can modify it to print all console logs.
   - personally  I like the idea to get a little info from the webpage and get the rest via the console 
- it's currently set to app.logger.info but I may change to debug.

File: src/emhass/web_server.py [hunk 1]
```diff
@@ -49,46 +102,64 @@ def action_call(action_name):
     input_data_dict = set_input_data_dict(config_path, str(data_path), costfun,
```
Comment (GeoDerp) [inline src/emhass/web_server.py:102]: There are a few places errors can occur if url or key is wrong, these lines are one. (then the other two are on ML Fit & Publish)

A suggested change for this line could be:
```python
    ActionStr = " >> Test HA connection..."
    app.logger.info(ActionStr)
    try:
        input_data_dict = set_input_data_dict(config_path, str(data_path), costfun, 
            params, runtimeparams, action_name, app.logger)
    except Exception as e :
        return make_response(grabLog(ActionStr), 400)
```

Comment (GeoDerp) [reply src/emhass/web_server.py:101]: The other method you could do is rap each action in a try and expect. Expect pasting exception in app.logger.error then returning make_response

**Comment by GeoDerp** [comment]
I'm hopping that this logger implementation makes sense and can help others send informative information back to the html.

**Comment by GeoDerp** [comment]
Sorry @davidusb-geek I closed the other PR as I merged those changes into this one as they are kinda related now.
Last things I added 
- url and key being incorect produces logger.error
- sensor being incorect produces logger.error
- days_to_retrieve default to 2 if number is under 2. It also produces  logger.warning

**Comment by GeoDerp** [comment]
@davidusb-geek  I'm sorry in advance for all the testing

**Comment by GeoDerp** [comment]
Just a heads up @davidusb-geek. The reason why i'm setting returns on the functions back to web_server, is that if there is an uncaught python error it stops the current function and runs the next in the action causing more errors (or just dosn't finish the function at all).
Whereas, if I return a string indicating an error has occurred in that function, I can stop that action and send back the log to the user on the website.

This is just how I would program it. Back when I did C in uni, we used to set returns on every function with 0 (fail) or 1 (pass). We could then check if that function worked every time and move to next. If not print issue.

**Comment by GeoDerp** [comment]
@davidusb-geek  Im pritty happy with this PR now. It's not perfect but I don't believe there will be any problems.
I'm currently looking now to see how difficult it would be to set True or Fail returns on every logger.error and end of each function.

**Comment by davidusb-geek** [comment]
Ok nice work!
I will at this and test it during the week-end

**Comment by GeoDerp** [comment]
If you're interested; https://github.com/GeoDerp/emhass/tree/Error-Suppression-Testing is my testing branch for testing returns on all functions and if statements to make sure those functions ran without an issue. *(If not, return back to webserver and the user will see the logger.errors)*
I literally just pasted responses and if statements everywhere without thinking about it methodically 😅. 
But it might be a good idea of what my concept is. 🤔

**Review by davidusb-geek — APPROVED** [review]

**Pull request merged**